server/routes/serviceProviderReferrals/interventionProgressPresenter.ts

```diff
@@ -53,18 +54,46 @@ export default class InterventionProgressPresenter {
     }
 
     return this.actionPlanAppointments.map(appointment => {
-      const editUrl = `/service-provider/action-plan/${this.actionPlan!.id}/sessions/${appointment.sessionNumber}/edit`
+      const sessionTableParams = this.sessionTableParams(appointment)
 
       return {
         sessionNumber: appointment.sessionNumber,
         appointmentTime: DateUtils.formatDateTimeOrEmptyString(appointment.appointmentTime),
-        tagArgs: appointment.appointmentTime
-          ? { text: 'SCHEDULED', classes: 'govuk-tag--blue' }
-          : { text: 'NOT SCHEDULED', classes: 'govuk-tag--grey' },
-        linkHtml: appointment.appointmentTime
-          ? `<a class="govuk-link" href="${editUrl}">Reschedule session</a><br><a class="govuk-link" href="/service-provider/action-plan/${this.actionPlan?.id}/appointment/${appointment.sessionNumber}/post-session-feedback/attendance">Give feedback</a>`
-          : `<a class="govuk-link" href="${editUrl}">Edit session details</a>`,
+        tagArgs: { text: sessionTableParams.text, classes: sessionTableParams.classes },
+        linkHtml: sessionTableParams.linkHTML,
       }
     })
   }
+
+  private sessionTableParams(appointment: ActionPlanAppointment): { text: string; classes: string; linkHTML: string } {
+    const sessionFeedbackAttendance = appointment.sessionFeedback.attendance
```

Comment:
As you say in your next commit, there's some heavy overlap between this and the PP page, but they differ in the HTML used for the links.

I think that there are perhaps three separate concerns, which this method is mixing:

1. Given an appointment, what is its status? (all the `if` statements here)
2. Given an appointment status, what text and what colour tag should be displayed?
3. Given an appointment status, what link should be displayed?

The first two should be shared between the SP and PP pages, the third possibly not.

So how about this?

In `sessionStatus.ts`:

```typescript
enum SessionStatus {
  notScheduled,
  scheduled,
  completed,
  didNotAttend,
}

export default {
  forAppointment: (appointment: ActionPlanAppointment) => {
    // all the if statements to determine the status from the appointment and return a `SessionStatus` value
  }
}
```

Then something to encapsulate the shared presentation logic:

```typescript
class SessionStatusPresenter {
  constructor(private readonly status: SessionStatus) {}

  get text(): string {
    switch (this.status) {
      case SessionStatus.notAttended:
        return 'not attended'
      // etc
    }
  }

  get tagClass(): string {
    switch (this.status) {
      case SessionStatus.notAttended:
        return 'govuk-tag--grey'
      // etc
    }
  }

}
```

And then in the presenter you'd do something like

```typescript
private sessionTableParams(appointment: ActionPlanAppointment): { text: string; classes: string; linkHTML: string } {
  const status = sessionStatus.forAppointment(appointment)
  const presenter = new SessionStatusPresenter(status)

  switch (status) {
    case SessionStatus.notAttended:
      return { ...presenter, linkHTML: `<a class="govuk-link" href="#">View feedback form</a>` }
    // etc
  }
}
```

Reply:
This was really helpful, and really satisfying - thanks for that. Have made these changes